File: router_bench.py
# -*- coding: utf-8 -*-

##
## See:
##      https://pythonhosted.org/Benchmarker/
## for details of 'benchmarker.py'.
##

import sys, os, re
from benchmarker import Benchmarker

from minikeight import (
    on, RequestHandler,
    NaiveLinearRouter, PrefixLinearRouter, FixedLinearRouter, HashedLinearRouter,
    NaiveRegexpRouter, SmartRegexpRouter, NestedRegexpRouter,
    OptimizedRegexpRouter, SlicedRegexpRouter, HashedRegexpRouter,
    TrieRouter, StateMachineRouter,
)


router_classes = (
    NaiveLinearRouter,
    PrefixLinearRouter,
    FixedLinearRouter,
    HashedLinearRouter,
    #
    NaiveRegexpRouter,
    SmartRegexpRouter,
    NestedRegexpRouter,
    OptimizedRegexpRouter,
    SlicedRegexpRouter,
    HashedRegexpRouter,
    #
    TrieRouter,
    StateMachineRouter,
)


class DummyAPI(RequestHandler):

    with on.path('/'):

        @on('GET')
        def do_index(self):
            return {"action": "index"}

        @on('POST')
        def do_create(self):
            return {"action": "create"}

    with on.path('/new'):

        @on('GET')
        def do_new(self, id):
            return {"action": "new", "id": id}

    #with on.path('/{id:int}.*'):
    with on.path('/{id:int}.json'):

        @on('GET')
        def do_show(self, id):
            return {"action": "show", "id": id}

        @on('PUT')
        def do_update(self, id):
            return {"action": "update", "id": id}

        @on('DELETE')
        def do_delete(self, id):
            return {"action": "delete", "id": id}

    with on.path('/{id:int}/edit'):

        @on('GET')
        def do_edit(self, id):
            return {"action": "edit", "id": id}


class CommentDummyAPI(RequestHandler):

    with on.path('/'):

        @on('GET')
        def do_index(self, parent_id):
            return {"action": "index", "parent_id": parent_id}

    #with on.path('/{comment_id:int}.*'):
    with on.path('/{comment_id:int}.json'):

        @on('GET')
        def do_show(self, parent_id, comment_id):
            return {"action": "show", "parent_id": parent_id, "comment_id": comment_id}


class BlablaAPI(RequestHandler):

    with on.path('/'):

        @on('GET')
        def do_index(self, parent_id, comment_id):
            return {"action": "show", "ids": [parent_id, comment_id]}

    #with on.path('/{blabla_id:int}.*'):
    with on.path('/{blabla_id:int}.json'):

        @on('GET')
        def do_show(self, parent_id, comment_id, blabla_id):
            return {"action": "show", "ids": [parent_id, comment_id, blabla_id]}


class MockAPI(RequestHandler):

    with on.path(''):

        @on('GET')
        def do_any(self, *params):
            return {"params": params}


benchtype = os.environ.get('BENCHTYPE') or None
if benchtype == None:

    urlpaths = (
        '/api/aaa/',
        '/api/zzz/',
        '/api/aaa/123.json',
        '/api/zzz/789.json',
        '/api/aaa/123/comments/999.json',
        '/api/zzz/789/comments/999.json',
        #'/api/aaa/123/comments/999/blabla/888.json',
        #'/api/zzz/789/comments/999/blabla/888.json',
    )

    import string
    dct = {}
    for c in string.ascii_lowercase:
        path = "/%s" % (c * 3)   # ex: "/aaa", "/bbb", ...
        dct[path] = DummyAPI
        dct[path+"/{id:int}/comments"] = CommentDummyAPI
        #dct[path+"/{id:int}/comments/{comment_id:int}/blabla"] = BlablaAPI
    mapping = {"/api": dct}
    ## same as below
    #mapping = {
    #    "/api": {
    #        {"/aaa": DummyAPI, "/aaa/{id:int}/comments": CommentDummyAPI},
    #        {"/bbb": DummyAPI, "/bbb/{id:int}/comments": CommentDummyAPI},
    #        {"/ccc": DummyAPI, "/ccc/{id:int}/comments": CommentDummyAPI},
    #        {"/ddd": DummyAPI, "/ddd/{id:int}/comments": CommentDummyAPI},
    #        {"/eee": DummyAPI, "/eee/{id:int}/comments": CommentDummyAPI},
    #        {"/fff": DummyAPI, "/fff/{id:int}/comments": CommentDummyAPI},
    #        {"/ggg": DummyAPI, "/ggg/{id:int}/comments": CommentDummyAPI},
    #        {"/hhh": DummyAPI, "/hhh/{id:int}/comments": CommentDummyAPI},
    #        {"/iii": DummyAPI, "/iii/{id:int}/comments": CommentDummyAPI},
    #        {"/jjj": DummyAPI, "/jjj/{id:int}/comments": CommentDummyAPI},
    #        {"/kkk": DummyAPI, "/kkk/{id:int}/comments": CommentDummyAPI},
    #        {"/lll": DummyAPI, "/lll/{id:int}/comments": CommentDummyAPI},
    #        {"/mmm": DummyAPI, "/mmm/{id:int}/comments": CommentDummyAPI},
    #        {"/nnn": DummyAPI, "/nnn/{id:int}/comments": CommentDummyAPI},
    #        {"/ooo": DummyAPI, "/ooo/{id:int}/comments": CommentDummyAPI},
    #        {"/ppp": DummyAPI, "/ppp/{id:int}/comments": CommentDummyAPI},
    #        {"/qqq": DummyAPI, "/qqq/{id:int}/comments": CommentDummyAPI},
    #        {"/rrr": DummyAPI, "/rrr/{id:int}/comments": CommentDummyAPI},
    #        {"/sss": DummyAPI, "/sss/{id:int}/comments": CommentDummyAPI},
    #        {"/ttt": DummyAPI, "/ttt/{id:int}/comments": CommentDummyAPI},
    #        {"/uuu": DummyAPI, "/uuu/{id:int}/comments": CommentDummyAPI},
    #        {"/vvv": DummyAPI, "/vvv/{id:int}/comments": CommentDummyAPI},
    #        {"/www": DummyAPI, "/www/{id:int}/comments": CommentDummyAPI},
    #        {"/xxx": DummyAPI, "/xxx/{id:int}/comments": CommentDummyAPI},
    #        {"/yyy": DummyAPI, "/yyy/{id:int}/comments": CommentDummyAPI},
    #        {"/zzz": DummyAPI, "/zzz/{id:int}/comments": CommentDummyAPI},
    #    },
    #}

elif benchtype == "many":

    urlpaths = (
        '/api/aaa000/',
        '/api/zzz099/',
        '/api/aaa000/123.json',
        '/api/zzz099/789.json',
        '/api/aaa000/123/comments/999.json',
        '/api/zzz099/789/comments/999.json',
        #'/api/aaa000/123/comments/999/blabla/888.json',
        #'/api/zzz099/789/comments/999/blabla/888.json',
    )

    import string
    dct = {}
    for i in range(100):
        for c in string.ascii_lowercase:
            path = "/%s" % (c * 3) + "%03d" % i
            dct[path] = DummyAPI
            dct[path+"/{id:int}/comments"] = CommentDummyAPI
            #dct[path+"/{id:int}/comments/{comment_id:int}/blabla"] = BlablaAPI
    mapping = {"/api": dct}

elif benchtype == "githubapi":

    urlpaths = [
        #'/repos/{owner}/{repo}',
        '/repos/owner1/repo2',
        #'/repos/{owner}/{repo}/pulls/{pull_number}/reviews/{review_id}/comments',
        '/repos/owner1/repo2/pulls/333/reviews/444/comments',
        #'/teams/{team_id}/memberships/{username}',
        '/teams/12345/memberships/username1',
    ]
    urlpaths = [ "/api" + x for x in urlpaths ]

    datafile = "data/github-api-paths.txt"
    dct = {}
    callback = lambda m: m.group(0).replace('-', '_')  # ex: '{foo-id}' -> '{foo_id}'
    with open(datafile) as f:
        for line in f:
            #urlpath = line.strip()
            urlpath = re.sub(r'\{.*\}', callback, line.strip())
            if not urlpath:
                continue
            if urlpath.startswith('#'):
                continue
            dct[urlpath] = MockAPI
    mapping = {"/api": dct}

else:

    raise Exception("%s : Unknonw benchmark type." % (benchtype,))


def validate(urlpath, result):
    if benchtype == None or benchtype == "many":
        if urlpath.endswith(('/123', '/123.json')):
            assert result == (DummyAPI, DummyAPI.do_show, [123]), "result=%r" % (result,)
        elif urlpath.endswith(('/789', '/789.json')):
            assert result == (DummyAPI, DummyAPI.do_show, [789]), "result=%r" % (result,)
        elif urlpath.endswith(('/123/comments/999', '/123/comments/999.json')):
            assert result == (CommentDummyAPI, CommentDummyAPI.do_show, [123, 999]), "result=%r" % (result,)
        elif urlpath.endswith(('/789/comments/999', '/789/comments/999.json')):
            assert result == (CommentDummyAPI, CommentDummyAPI.do_show, [789, 999]), "result=%r" % (result,)
        elif urlpath.endswith(('/blabla/', '/blabla.json')):
            assert result[0:2] == (BlablaAPI, BlablaAPI.do_index), "result=%r" % (result,)
        elif urlpath.endswith(('/blabla/888', '/blabla/888.json')):
            assert result[0:2] == (BlablaAPI, BlablaAPI.do_show), "result=%r" % (result,)
        else:
            assert result == (DummyAPI, DummyAPI.do_index, []), "result=%r" % (result,)
    elif benchtype == "githubapi":
        if urlpath == "/api/repos/owner1/repo2":
            assert result == (MockAPI, MockAPI.do_any, ["owner1", "repo2"]), "result=%r" % (result,)
        elif urlpath == "/api/repos/owner1/repo2/pulls/333/reviews/444/comments":
            assert result == (MockAPI, MockAPI.do_any, ["owner1", "repo2", "333", "444"]), "result=%r" % (result,)
        elif urlpath == "/api/teams/12345/memberships/username1":
            assert result == (MockAPI, MockAPI.do_any, ["12345", "username1"]), "result=%r" % (result,)
        else:
            assert False, "urlpath=%r" % (urlpath,)
    else:
        raise Exception("%s : Unknonw benchmark type." % (benchtype,))


loop = int(os.environ.get('N') or 1000 * 1000)
width = 17 + max( len(x) for x in urlpaths )
with Benchmarker(loop, width=width, cycle=1, extra=0) as bench:

    debug = bench.properties.get('debug', False)

    for router_class in router_classes:
        for urlpath in urlpaths:
            label = router_class.__name__.replace('Router', '')
            if router_class.__name__.startswith("Hashed"):
                router_obj = router_class(mapping, r'^/api/\w\w')
            else:
                router_obj = router_class(mapping)

            @bench("%-15s: %-16s" % (label, urlpath))
            def _(bm, router=router_obj, urlpath=urlpath):
                for _ in bm:
                    result = router.lookup('GET', urlpath)
                #
                if debug:
                    validate(urlpath, result)

File: router_bench.rb
# -*- coding: utf-8 -*-
# frozen_string_literal: true

require 'benchmarker'  # https://kwatch.github.io/benchmarker/ruby.html

require './minikeight'

router_classes = [
  K8::NaiveLinearRouter,
  K8::PrefixLinearRouter,
  K8::FixedLinearRouter,
  K8::HashedLinearRouter,
  #
  K8::NaiveRegexpRouter,
  K8::SmartRegexpRouter,
  K8::NestedRegexpRouter,
  K8::OptimizedRegexpRouter,
  K8::SlicedRegexpRouter,
  K8::HashedRegexpRouter,
 #K8::HashedRegexpRouter1,
 #K8::HashedRegexpRouter2,
  #
  K8::ListBasedTrieRouter,
  K8::DictBasedTrieRouter,
  K8::StateMachineRouter,
  K8::NestedDictRouter,
]


class HelloAPI < K8::RequestHandler
  mapping "/"         , :GET=>:do_list, :POST=>:do_create
  mapping "/new"      , :GET=>:do_new
  mapping "/{id}.json", :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete
  mapping "/{id}/edit", :GET=>:do_edit
  def do_list()     ; return {type: "list"}           ; end
  def do_create()   ; return {type: "create"}         ; end
  def do_new()      ; return {type: "new"}            ; end
  def do_show(id)   ; return {type: "show"  , id: id} ; end
  def do_update(id) ; return {type: "update", id: id} ; end
  def do_delete(id) ; return {type: "delete", id: id} ; end
  def do_edit(id)   ; return {type: "edit"  , id: id} ; end
end

class CommentAPI < K8::RequestHandler
  mapping "/"         , :GET=>:do_list, :POST=>:do_create
  mapping "/new"      , :GET=>:do_new
  mapping "/{id}.json", :GET=>:do_show, :PUT=>:do_update, :DELETE=>:do_delete
  mapping "/{id}/edit", :GET=>:do_edit
  def do_list(id)         ; return {type: "list"}           ; end
  def do_create(id)       ; return {type: "create"}         ; end
  def do_new(id)          ; return {type: "new"}            ; end
  def do_show(id, c_id)   ; return {type: "show"  , id: id} ; end
  def do_update(id, c_id) ; return {type: "update", id: id} ; end
  def do_delete(id, c_id) ; return {type: "delete", id: id} ; end
  def do_edit(id, c_id)   ; return {type: "edit"  , id: id} ; end
end

class DummyAPI < K8::RequestHandler
  mapping "", :GET=>:do_nothing
  def do_nothing(*args)
    return {args: args}
  end
end

benchdata = ENV['BENCHDATA']
case benchdata
when nil, "", "default"

  mapping = {
    "/api" => {
      "/aaa" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/bbb" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ccc" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ddd" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/eee" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/fff" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ggg" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/hhh" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/iii" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/jjj" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/kkk" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/lll" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/mmm" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/nnn" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ooo" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ppp" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/qqq" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/rrr" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/sss" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/ttt" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/uuu" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/vvv" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/www" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/xxx" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/yyy" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
      "/zzz" => {"" => HelloAPI, "/{id}/comments" => CommentAPI},
    }
  }

  request_paths = [
    #"/api/aaa/",
    "/api/aaa/123.json",
    "/api/aaa/123/comments/888.json",
    #"/api/zzz/",
    "/api/zzz/456.json",
    "/api/zzz/456/comments/999.json",
  ]

  width = 52

when "githubapi"

  dict = {}
  File.open("data/github-api-paths.txt") do |f|
    f.each do |line|
      line = line.strip
      next if line.empty?
      next if line =~ /^#/
      urlpath_pat = line.gsub(/\{(.*?)\}/) { "{#{$1.gsub('-', '_')}}" }
      dict[urlpath_pat] = DummyAPI
    end
  end
  mapping = {"/api" => dict}

  request_paths = [
    #'/gists/{gist_id}/star',
    '/gists/1234567890/star',
    #'/repos/{owner}/{repo}',
    '/repos/owner1/repo2',
    #'/repos/{owner}/{repo}/pulls/{pull_number}/reviews/{review_id}/comments',
    '/repos/owner1/repo2/pulls/333/reviews/444/comments',
    #'/teams/{team_id}/memberships/{username}',
    '/teams/12345/memberships/username1',
  ].collect {|x| "/api" + x }

  width = 75

else

  raise "#{benchdata}: unknown bench type."

end


N = Integer(ENV['N'] || 1000_000)

title = "router benchmark"
Benchmarker.scope(title, width: width, loop: 1, iter: nil, extra: nil) do
  ## other options -- inverse: true, outfile: "result.json", quiet: true,
  ##                  sleep: 1, colorize: true, filter: "task=*foo*"

  ## tasks
  task nil do    # empty-loop task
    n = N
    i = 0
    while (i += 1) <= n
      # do nothing
    end
  end

  router_classes.each do |router_class|
    request_paths.each do |request_path|
      #
      proc {|klass, req_path|
        router = klass.new(mapping)
        class_name = klass.name.sub(/Router/, '').sub(/^K8::/, '')
        label = "%-15s: %-16s" % [class_name, req_path]
        task label do
          req_path_ = req_path
          router_   = router
          n = N
          i = 0
          while (i += 1) <= n
            result = router_.find(req_path_)
          end
          result
        end
      }.call(router_class, request_path)
      #
    end
  end

  ## validation
  validate do |val, title|   # or: validate do |val, task_name, tag|
    case title.strip
    ## benchdata: "default"
    when /\/api\/aaa\/$/, /\/api\/zzz\/$/
      expected = [HelloAPI, {GET: :do_list, POST: :do_create}, []]
      assert_eq val, expected
    when /\/123\.json$/
      expected = [HelloAPI, {GET: :do_show, PUT: :do_update, DELETE: :do_delete}, [123]]
      assert_eq val, expected
    when /\/456\.json$/
      expected = [HelloAPI, {GET: :do_show, PUT: :do_update, DELETE: :do_delete}, [456]]
      assert_eq val, expected
    when /\/comments\/888\.json$/
      expected = [CommentAPI, {GET: :do_show, PUT: :do_update, DELETE: :do_delete}, [123, 888]]
      assert_eq val, expected
    when /\/comments\/999\.json$/
      expected = [CommentAPI, {GET: :do_show, PUT: :do_update, DELETE: :do_delete}, [456, 999]]
      assert_eq val, expected
    ## benchdata: "githubapi"
    when %r!/gists/1234567890/star$!
      expected = [DummyAPI, {GET: :do_nothing}, [1234567890]]
      assert_eq val, expected
    when %r!/api/repos/owner1/repo2$!
      expected = [DummyAPI, {GET: :do_nothing}, ["owner1", "repo2"]]
      assert_eq val, expected
    when %r!/api/repos/owner1/repo2/pulls/333/reviews/444/comments$!
      expected = [DummyAPI, {GET: :do_nothing}, ["owner1", "repo2", "333", 444]]
      assert_eq val, expected
    when %r!/api/teams/12345/memberships/username1$!
      expected = [DummyAPI, {GET: :do_nothing}, [12345, "username1"]]
      assert_eq val, expected
    else
      puts "** interal error: title=#{title.inspect}"
    end
  end

end
